src/common/buffer.hpp

```diff
@@ -284,104 +284,85 @@ class RxCounters {
+//  */
+class FrameCounters {
 public:
     size_t max_symbol_count;
-    bool last_symbol(int frame_id)
-    {
-        const size_t frame_slot = frame_id % TASK_BUFFER_FRAME_NUM;
-        if (++symbol_count[frame_slot] == max_symbol_count) {
-            symbol_count[frame_slot] = 0;
-            return true;
-        }
-        return false;
-    }
+    size_t max_task_count;
 
-    void init(int _max_symbol_count)
+    void init(
+        int max_symbol_count, int max_task_count = -1, int max_data_symbol = -1)
     {
+        this->max_symbol_count = max_symbol_count;
+        this->max_task_count = max_task_count;
         symbol_count.fill(0);
-        max_symbol_count = _max_symbol_count;
+        //define task counters if #tasks is > 0
+        if (max_task_count > 0) {
+            if (max_data_symbol <= 0)
+                throw "the max data symbol should be greater than zero for a "
+                      "positive max task count.";
+
+            for (size_t i = 0; i < TASK_BUFFER_FRAME_NUM; i++)
+                task_count[i] = new size_t[max_data_symbol]();
+        }
     }
 
-    size_t get_symbol_count(size_t frame_id)
+    void fini()
     {
-        return symbol_count[frame_id % TASK_BUFFER_FRAME_NUM];
+        for (size_t i = 0; i < TASK_BUFFER_FRAME_NUM; i++)
+            delete[] task_count[i];
     }
 
-private:
-    std::array<size_t, TASK_BUFFER_FRAME_NUM> symbol_count;
-};
-
-class ZF_stats : public Frame_stats {
-public:
-    size_t coded_frame;
-    size_t& max_task_count;
-    ZF_stats(void)
-        : max_task_count(max_symbol_count)
-    {
-    }
-    void init(int max_tasks)
+    /**
+     * Checks whether or not the symbol is the last symbol for a given frame while simultaneously incrementing the symbol count.
+     * If the symbol is the last symbol, the count is reset to 0.
+     * @arg frame id: the frame id to check
+     **/
+    bool last_symbol(int frame_id)
     {
-        Frame_stats::init(max_tasks);
-        coded_frame = SIZE_MAX;
+        const size_t frame_slot = frame_id % TASK_BUFFER_FRAME_NUM;
+        if (++symbol_count[frame_slot] == max_symbol_count) {
+            symbol_count[frame_slot] = 0;
+            return true;
+        }
+        return false;
     }
-};
-
-class Data_stats : public Frame_stats {
-public:
-    size_t max_task_count;
 
-    void init(int _max_task_count, int max_symbols, int max_data_symbol)
-    {
-        Frame_stats::init(max_symbols);
-        for (size_t i = 0; i < TASK_BUFFER_FRAME_NUM; i++)
-            task_count[i] = new size_t[max_data_symbol]();
-        max_task_count = _max_task_count;
-    }
-    void fini()
-    {
-        for (size_t i = 0; i < TASK_BUFFER_FRAME_NUM; i++)
-            delete[] task_count[i];
-    }
-    bool last_task(size_t frame_id, size_t data_symbol_id)
+    /**
+     * Checks whether or not the task is the last task for a given frame and symbol while simultaneously incrementing the task count.
```

Comment:
We need to break down this long line into two to keep the column limit as 80 following the .clang-format file.